packages/block-editor/src/hooks/utils.js

```diff
@@ -30,8 +30,75 @@ export const cleanEmptyObject = ( object ) => {
 	return isEmpty( cleanedNestedObjects ) ? undefined : cleanedNestedObjects;
 };
 
+/**
+ * Converts a path to an array of its fragments.
+ * Supports strings, numbers and arrays:
+ *
+ * 'foo' => [ 'foo' ]
+ * 'foo.bar.baz' => [ 'foo', 'bar', 'baz' ]
+ * 2 => [ '2' ]
+ * [ 'foo', 'bar' ] => [ 'foo', 'bar' ]
+ *
+ * @param {number|string|Array} path Path
+ * @return {Array} Normalized path.
+ */
+function normalizePath( path ) {
+	if ( Array.isArray( path ) ) {
+		return path;
+	} else if ( typeof path === 'number' ) {
+		return [ path.toString() ];
+	}
+
+	return path.split( '.' ).filter( Boolean );
+}
+
+/**
+ * Clones an object.
+ * Non-object values are returned unchanged.
+ *
+ * @param {*} object Object to clone.
+ * @return {*} Cloned object, or original literal non-object value.
+ */
+function cloneObject( object ) {
+	if ( typeof object === 'object' ) {
+		return {
+			...Object.fromEntries(
+				Object.entries( object ).map( ( [ key, value ] ) => [
+					key,
+					cloneObject( value ),
+				] )
+			),
+		};
+	}
+
+	return object;
+}
+
+/**
+ * Perform an immutable set.
+ * Handles nullish initial values.
+ * Clones all nested objects in the specified object.
+ *
+ * @param {Object}              object Object to set a value in.
+ * @param {number|string|Array} path   Path in the object to modify.
```

Comment:
I would prefer to see the string-based API removed altogether; parsing strings to split them into paths is quite inefficient if it ends up in a hot path. I've had to fix several instances of this in Redux selectors over the years (in other applications), and the fix was usually to move to a (static) array path instead.

There's also the matter of this API being more complex and less explicit, which makes it difficult to migrate to something simpler in the future, even if the number and string notations aren't being used. If at this time it's easy to identify all the usage and move to something simpler, then we should take the chance, lest it become more difficult to do so in the future.

The only argument in favour of the dot notation, as I see it, is readability, but I don't think it's worth the other issues, given that an array of path segments is pretty readable on its own. Plus, if we really wanted to keep dotted paths for some reason, we could always relegate that to the consumer, where it makes sense:

```js
import { immutableSet } from 'hooks/utils';

const THE_PATH = 'foo.bar.baz'.split( '.' );

function doSomething(input) {
  immutableSet( input, THE_PATH, 42 );
}
```

The above avoids parsing the path and instancing an array on every call, and is just as readable as passing it directly to `immutableSet`.

All of that said, I acknowledge this is possibly out of scope for this PR.

Reply:
Thank you for the observation and your hint, @sgomes! You actually caught one of my biggest doubts about this work here. 

I was very close to removing the `.` support, and likely should have, since we have no present usage of it in the codebase. That will simplify the code and add less complexity to an already complex part of the codebase. 

I consider this part of the PR's scope, since we're directly tinkering with the API, so it's the right time to simplify it.

I'll work on that before asking you to do another review.